### PR TITLE
Feature 3: updates to jsonapi module to handle revisions

### PR DESCRIPTION
**Pull request by justageek**

### CODE REVIEW

**Comment by q0rban** [comment]
@justageek is this a patch to the module? If so, you should use composer-patches and link directly to the Drupal.org patch file in composer.json. If not, hacking the JSONApi module directly will block future upgrades.

**Comment by justageek** [comment]
@q0rban , I'm currently just trying to get something work, this is for review, when I know this is ok I will create the patch in a different branch and apply it as mentioned.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 1]
```diff
@@ -132,6 +132,45 @@ public function getIndividual(EntityInterface $entity, Request $request, $respon
...
+   *   The revision id or one of 'current' or 'latest'.
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request object.
+   * @param int $response_code
+   *   The response code. Defaults to 200.
+   *
+   * @return \Drupal\jsonapi\ResourceResponse
+   *   The response.
+   */
+  public function getRevision(EntityInterface $entity, Request $request, $response_code = 200) {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: I think we can drop the `$response_code` parameter.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 2]
```diff
@@ -132,6 +132,45 @@ public function getIndividual(EntityInterface $entity, Request $request, $respon
...
+    $storage = $this->entityTypeManager->getStorage($this->resourceType->getEntityTypeId());
+
+    if ($revision_id === 'latest') {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: `$revision_id` is documented as a method parameter, but it is not in the parameters declaration.

Comment (e0ipso) [reply web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: Also, let's avoid magic numbers like `'latest'`. Let's move it to the class constants.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 3]
```diff
@@ -132,6 +132,45 @@ public function getIndividual(EntityInterface $entity, Request $request, $respon
...
+      if ($revision_id = $storage->getLatestRevisionId($entity->id())) {
+        $entity = $storage->loadRevision($revision_id);
+      }
+    }
+    elseif (is_numeric($revision_id) && $entity->getLoadedRevisionId() != $revision_id) {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: We should not be expecting revision IDs but version UUIDs instead. This means that we can't detect an explicit revision by checking `is_numeric`. Instead we can check if it's not `static::LATEST`, `static::CURRENT` and it is `Drupal\Component\Uuid::isValid($revision_uuid)`.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 4]
```diff
@@ -132,6 +132,45 @@ public function getIndividual(EntityInterface $entity, Request $request, $respon
     return $response;
   }
 
+  /**
+   * Gets an individual entity revision.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The loaded entity.
+   * @param int|string $revision_id
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: We probably want something like:

```
@param string $revision_uuid
```

instead.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 5]
```diff
@@ -340,13 +379,19 @@ public function deleteIndividual(EntityInterface $entity, Request $request) {
    * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
    */
   public function getCollection(Request $request) {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: Let's hold off on this for now. I think there is another issue to deal with this part.

File: web/modules/contrib/jsonapi/src/Controller/RequestHandler.php [hunk 6]
```diff
@@ -210,6 +217,9 @@ protected function action(Request $request, ResourceType $resource_type) {
         elseif ($request->get('related')) {
           return 'getRelated';
         }
+        elseif ($request->get('revision_id')) {
+          return $request->get($resource_type->getEntityTypeId()) ? 'getRevision' : 'getCollection';
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/RequestHandler.php]: Nit: rename to `getIndividualRevision`.

File: web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php [hunk 7]
```diff
@@ -202,17 +214,18 @@ public function isLocatable() {
...
-   *   The deserialization target class.
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type.
    * @param bool $internal
    *   (optional) Whether the resource type should be internal.
    * @param bool $is_locatable
    *   (optional) Whether the resource type is locatable.
    */
-  public function __construct($entity_type_id, $bundle, $deserialization_target_class, $internal = FALSE, $is_locatable = TRUE) {
+  public function __construct($entity_type_id, $bundle, EntityTypeInterface $entity_type, $internal = FALSE, $is_locatable = TRUE) {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php]: I understand the motivation, but changing the signature like this is a breaking change for `1.x`. Can we slate this change for `2.x` only?

File: web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php [hunk 8]
```diff
@@ -68,6 +70,16 @@ public function getEntityTypeId() {
     return $this->entityTypeId;
   }
 
+  /**
+   * Gets the entity type.
+   *
+   * @return \Drupal\Core\Entity\EntityTypeInterface
+   *   The entity type.
+   */
+  public function getEntityType() {
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php]: Let's keep the `ResourceType` class Drupal entity agnostic. Maybe we can try with `isVersionable()`.

File: web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php (hunk 7 above)
Comment (justageek) [reply web/modules/contrib/jsonapi/src/ResourceType/ResourceType.php]: This is going to make it tough to get to the necessary service needed to determine when an entity type is revisionable, I will try to find a fallback to some other class / interface check instead.

File: web/modules/contrib/jsonapi/src/Controller/EntityResource.php [hunk 9]
```diff
@@ -84,6 +84,20 @@ class EntityResource {
...
+   * @var string
+   */
+  const LATEST = 'latest';
+
+  /**
+   * Revision ID value representing the current revsion.
+   *
+   * @var string
+   */
+  const CURRENT = 'current';
```
Comment (e0ipso) [inline web/modules/contrib/jsonapi/src/Controller/EntityResource.php]: Let's rename to `'working-copy'`

**Comment by e0ipso** [comment]
This is how it works after the changes:

https://monosnap.com/file/qEUy4a4Im1SEToAZVcpnicCDqqpRSa

**Comment by sethlbrown** [comment]
@e0ipso do we still need this PR open?

**Pull request closed without merging**